File: tui/legal_term/live_client.py
"""
LiveClient — connects to a running legal-mcp SSE server.

Usage:
  Set  LEGAL_MCP_URL=http://localhost:8000  in your environment (or a .env
  file loaded before starting the TUI), then run:

      legal-term --live

  Or import and pass directly:
      from legal_term.live_client import LiveClient
      app = LegalTermApp(client=LiveClient())

The server must already be running:
    cd legal-mcp && pip install -e . && python main.py --transport sse
"""
from __future__ import annotations

import json
from typing import Any

from mcp import ClientSession
from mcp.client.sse import sse_client

from .client import (
    AnalysisJob, AuditEntry, Case, CitationResult, Contract, ContractClause,
    LegalMcpClient, PrivilegeResult, Statute, Workflow, WorkflowStep,
)

FIXTURES_WORKFLOWS = None  # lazy-loaded fallback for get_workflows


async def _call(session: ClientSession, name: str, **kwargs: Any) -> Any:
    """Call an MCP tool and parse the JSON text result."""
    result = await session.call_tool(name, arguments=kwargs)
    for part in result.content:
        if hasattr(part, "text"):
            return json.loads(part.text)
    raise RuntimeError(f"Tool {name!r} returned no text content")


# ─── normalisers ─────────────────────────────────────────────────────────────

def _case(d: dict[str, Any]) -> Case:
    return Case(
        id=d.get("id", ""),
        name=d.get("name", d.get("caseName", "")),
        citation=d.get("citation", ""),
        court=d.get("court", ""),
        jurisdiction=d.get("jurisdiction", ""),
        year=int(d.get("year", 0)),
        topics=d.get("topics", []),
        holding=d.get("holding", ""),
        summary=d.get("summary", ""),
        disposition=d.get("disposition", ""),
        relevance_score=float(d.get("relevance_score", 0.5)),
        cites=d.get("cites", []),
    )


def _statute(d: dict[str, Any]) -> Statute:
    return Statute(
        id=d.get("id", ""),
        title=d.get("title", ""),
        jurisdiction=d.get("jurisdiction", ""),
        citation=d.get("citation", ""),
        text=d.get("text", ""),
        enacted=d.get("enacted", ""),
        last_amended=d.get("last_amended", ""),
        history=d.get("history", ""),
        topics=d.get("topics", []),
    )


def _clause(d: dict[str, Any]) -> ContractClause:
    return ContractClause(
        key=d.get("key", d.get("clause_type", "")),
        label=d.get("label", d.get("name", d.get("clause_type", ""))),
        text=d.get("text", d.get("content", "")),
        risk=str(d.get("risk", d.get("risk_level", "MEDIUM"))).upper(),
        risk_note=d.get("risk_note", d.get("note", d.get("rationale", ""))),
    )


def _contract(contract_id: str, d: dict[str, Any]) -> Contract:
    raw_clauses = d.get("clauses", [])
    clauses = [_clause(c) for c in raw_clauses] if isinstance(raw_clauses, list) else []
    return Contract(
        id=contract_id,
        title=d.get("title", d.get("contract_id", contract_id)),
        type=d.get("type", d.get("contract_type", "Unknown")),
        risk_level=str(d.get("risk_level", "MEDIUM")).upper(),
        clauses=clauses,
        missing_clauses=d.get("missing_clauses", []),
        parties=d.get("parties", {}),
        governing_law=d.get("governing_law", ""),
        term=d.get("term", ""),
        liability_cap=d.get("liability_cap"),
    )


def _cite_result(citation: str, d: dict[str, Any]) -> CitationResult:
    return CitationResult(
        citation=citation,
        valid=bool(d.get("valid", d.get("is_valid", True))),
        normalized=str(d.get("normalized", d.get("normal", citation))),
        reporter=str(d.get("reporter", "")),
        reporter_name=str(d.get("reporter_name", d.get("reporter", ""))),
        issues=list(d.get("issues", d.get("errors", []))),
        integrity=str(d.get("integrity", "not_found" if not d.get("in_database") else "verified")),
    )


def _job(d: dict[str, Any]) -> AnalysisJob:
    return AnalysisJob(
        id=str(d.get("job_id", d.get("id", ""))),
        file=str(d.get("file_path", d.get("file", ""))),
        status=str(d.get("status", "queued")),
        queued_at=str(d.get("created_at", d.get("queued_at", ""))),
        completed_at=d.get("updated_at", d.get("completed_at")),
        risk_level=d.get("risk_level"),
        clause_count=d.get("clause_count"),
        flags=d.get("flags"),
        error=d.get("error"),
    )


# ─── LiveClient ──────────────────────────────────────────────────────────────

class LiveClient:
    """LegalMcpClient implementation that calls the live SSE server."""

    def __init__(self, url: str = "http://localhost:8000") -> None:
        self._url = url

    # Every public method opens a fresh SSE connection for simplicity.
    # For production, maintain a long-lived session instead.

    async def _session(self):  # type: ignore[return]
        """Async context manager yielding an active ClientSession."""
        async with sse_client(f"{self._url}/sse") as (read, write):
            async with ClientSession(read, write) as session:
                await session.initialize()
                yield session

    async def search_precedents(self, query: str) -> list[Case]:
        async with self._session() as s:
            data = await _call(s, "search_precedents", query=query)
            return [_case(c) for c in data.get("results", [])]

    async def extract_statute(self, statute_id: str) -> Statute | None:
        async with self._session() as s:
            data = await _call(s, "extract_statute", statute_id=statute_id)
            return None if data.get("error") else _statute(data)

    async def validate_citation(self, citation: str) -> CitationResult:
        async with self._session() as s:
            data = await _call(s, "validate_citation", citation=citation)
            return _cite_result(citation, data)

    async def normalize_citation(self, citation: str) -> CitationResult:
        async with self._session() as s:
            data = await _call(s, "normalize_citation", citation=citation)
            return _cite_result(citation, data)

    async def verify_citation_integrity(self, citation: str) -> CitationResult:
        async with self._session() as s:
            data = await _call(s, "verify_citation_integrity", citation=citation)
            return _cite_result(citation, data)

    async def analyze_contract(self, contract_id: str) -> Contract | None:
        try:
            async with self._session() as s:
                data = await _call(s, "analyze_clauses", contract_id=contract_id)
                return _contract(contract_id, data)
        except Exception:
            return None

    async def list_contracts(self) -> list[Contract]:
        return []  # no list-all tool on the server

    async def get_analysis_jobs(self) -> list[AnalysisJob]:
        async with self._session() as s:
            data = await _call(s, "list_analysis_jobs")
            return [_job(j) for j in data.get("jobs", [])]

    async def queue_document_analysis(self, file: str) -> AnalysisJob:
        async with self._session() as s:
            data = await _call(s, "queue_document_analysis", file_path=file)
            return AnalysisJob(
                id=str(data.get("job_id", "")),
                file=file,
                status="queued",
                queued_at=str(data.get("created_at", "")),
                completed_at=None,
                risk_level=None,
                clause_count=None,
                flags=None,
            )

    async def check_privilege_risk(self, _file: str, provider: str) -> PrivilegeResult:
        async with self._session() as s:
            data = await _call(s, "check_privilege_risk", provider=provider)
            risk = str(data.get("risk_level", "HIGH")).upper()
            return PrivilegeResult(
                risk=risk,
                label=f"{risk} — {provider}",
                provider=provider,
                indicators=data.get("privilege_indicators", []),
                recommendation=str(data.get("recommendation", "")),
                notice=str(data.get("notice", "")),
            )

    async def get_workflows(self) -> list[Workflow]:
        # Server has no list-all-workflows tool; fall back to fixtures.
        import json
        from pathlib import Path
        p = Path(__file__).parent.parent.parent / "fixtures" / "workflows.json"
        raw = json.loads(p.read_text(encoding="utf-8")) if p.exists() else []
        return [
            Workflow(
                id=w["id"], mnemonic=w["mnemonic"], title=w["title"],
                trigger=w["trigger"],
                steps=[WorkflowStep(**s) for s in w["steps"]],
            )
            for w in raw
        ]

    async def get_audit_log(self) -> list[AuditEntry]:
        # Audit log is local-only on the server side; not exposed via tool.
        return []


# Confirm the class satisfies the protocol
assert isinstance(LiveClient(), LegalMcpClient), "LiveClient does not satisfy LegalMcpClient protocol"

File: webterm/src/mcp/liveClient.ts
/**
 * LiveClient — wires the webterm to a running legal-mcp server via the
 * official MCP TypeScript SDK over SSE (default: http://localhost:8000/sse).
 *
 * Usage: set  VITE_MCP_URL=http://localhost:8000  in a .env.local file,
 * then swap the exported `client` singleton in this file for the one in
 * mockClient.ts inside src/mcp/index.ts.
 *
 * The server must be running:
 *   cd legal-mcp && pip install -e . && python main.py --transport sse
 */

import { Client } from '@modelcontextprotocol/sdk/client/index.js'
import { SSEClientTransport } from '@modelcontextprotocol/sdk/client/sse.js'
import type { LegalMcpClient } from './client'
import type {
  Case, Statute, Contract, ContractClause,
  AnalysisJob, AuditEntry, Workflow,
  CitationResult, PrivilegeResult, BriefOutline,
  NegotiationGuide, IntegrationStatus,
} from './types'

// ─── helpers ────────────────────────────────────────────────────────────────

function parse<T>(raw: unknown): T {
  if (typeof raw === 'string') return JSON.parse(raw) as T
  return raw as T
}

async function call<T>(c: Client, name: string, args: Record<string, unknown> = {}): Promise<T> {
  const result = await c.callTool({ name, arguments: args })
  const content = (result.content as { type: string; text?: string }[])[0]
  if (!content || content.type !== 'text' || !content.text) {
    throw new Error(`Tool ${name} returned no text content`)
  }
  return parse<T>(content.text)
}

// ─── client ─────────────────────────────────────────────────────────────────

export class LiveClient implements LegalMcpClient {
  private _client: Client | null = null
  private _url: string

  constructor(url = import.meta.env.VITE_MCP_URL ?? 'http://localhost:8000') {
    this._url = url
  }

  private async _connect(): Promise<Client> {
    if (this._client) return this._client
    const transport = new SSEClientTransport(new URL(`${this._url}/sse`))
    const client = new Client({ name: 'legal-terminal-webterm', version: '0.1.0' })
    await client.connect(transport)
    this._client = client
    return client
  }

  async searchPrecedents(query: string): Promise<Case[]> {
    const c = await this._connect()
    const data = await call<{ results: Case[] }>(c, 'search_precedents', { query })
    return data.results ?? []
  }

  async extractStatute(statuteId: string): Promise<Statute | null> {
    const c = await this._connect()
    const data = await call<Statute & { error?: string }>(c, 'extract_statute', { statute_id: statuteId })
    return data.error ? null : data
  }

  async validateCitation(citation: string): Promise<CitationResult> {
    const c = await this._connect()
    const data = await call<Record<string, unknown>>(c, 'validate_citation', { citation })
    return _normaliseCiteResult(citation, data)
  }

  async normalizeCitation(citation: string): Promise<CitationResult> {
    const c = await this._connect()
    const data = await call<Record<string, unknown>>(c, 'normalize_citation', { citation })
    return _normaliseCiteResult(citation, data)
  }

  async verifyCitationIntegrity(citation: string): Promise<CitationResult> {
    const c = await this._connect()
    const data = await call<Record<string, unknown>>(c, 'verify_citation_integrity', { citation })
    return _normaliseCiteResult(citation, data)
  }

  async analyzeContract(contractId: string): Promise<Contract | null> {
    const c = await this._connect()
    try {
      const data = await call<Record<string, unknown>>(c, 'analyze_clauses', { contract_id: contractId })
      return _normaliseContract(contractId, data)
    } catch {
      return null
    }
  }

  async searchCaseLaw(query: string): Promise<Case[]> {
    return this.searchPrecedents(query)
  }

  async researchLegalIssue(issue: string): Promise<{ cases: Case[]; statutes: Statute[]; summary: string }> {
    const c = await this._connect()
    try {
      const data = await call<{ cases?: Case[]; statutes?: Statute[]; summary?: string }>(
        c, 'research_legal_issue', { issue }
      )
      return { cases: data.cases ?? [], statutes: data.statutes ?? [], summary: data.summary ?? '' }
    } catch {
      const cases = await this.searchPrecedents(issue)
      return { cases, statutes: [], summary: '' }
    }
  }

  async listContracts(): Promise<Contract[]> {
    return []  // no dedicated list-all tool on the server
  }

  async compareContracts(id1: string, id2: string): Promise<{ base: Contract; compare: Contract; diffs: { key: string; change: string }[] }> {
    const c = await this._connect()
    const [base, cmp] = await Promise.all([
      this.analyzeContract(id1),
      this.analyzeContract(id2),
    ])
    const baseContract = base ?? _emptyContract(id1)
    const cmpContract = cmp ?? _emptyContract(id2)
    try {
      const data = await call<Record<string, unknown>>(c, 'compare_contracts', { contract_id_1: id1, contract_id_2: id2 })
      const diffs = (data.differences as { clause: string; change: string }[] | undefined) ?? []
      return { base: baseContract, compare: cmpContract, diffs: diffs.map(d => ({ key: d.clause ?? '', change: d.change ?? '' })) }
    } catch {
      return { base: baseContract, compare: cmpContract, diffs: [] }
    }
  }

  async suggestAlternatives(_clauseText: string, clauseType: string): Promise<string[]> {
    const c = await this._connect()
    const data = await call<{ alternatives: { text: string }[] }>(c, 'suggest_clause_alternatives', { clause_type: clauseType })
    return (data.alternatives ?? []).map(a => a.text)
  }

  async generateNegotiationGuide(contractId: string, partyRole: string): Promise<NegotiationGuide> {
    const c = await this._connect()
    const data = await call<NegotiationGuide>(c, 'generate_negotiation_guide', { contract_id: contractId, party_role: partyRole })
    return data
  }

  async analyzeDocument(file: string): Promise<{ metadata: Record<string, string | null>; risk_level: string; clauses: Contract['clauses'] }> {
    const c = await this._connect()
    const data = await call<Record<string, unknown>>(c, 'analyze_document', { file_path: file })
    const contract = _normaliseContract(file, data)
    return {
      metadata: { title: contract.title, type: contract.type, governing_law: contract.governing_law ?? null },
      risk_level: contract.risk_level,
      clauses: contract.clauses,
    }
  }

  async queueDocumentAnalysis(file: string): Promise<AnalysisJob> {
    const c = await this._connect()
    const data = await call<{ job_id: string; status: string; created_at: string }>(c, 'queue_document_analysis', { file_path: file })
    return {
      id: data.job_id,
      file,
      status: 'queued',
      queued_at: data.created_at,
      completed_at: null,
      risk_level: null,
      clause_count: null,
      flags: null,
    }
  }

  async getAnalysisJobs(): Promise<AnalysisJob[]> {
    const c = await this._connect()
    const data = await call<{ jobs: Record<string, unknown>[] }>(c, 'list_analysis_jobs', {})
    return (data.jobs ?? []).map(_normaliseJob)
  }

  async checkPrivilegeRisk(_file: string, provider: string): Promise<PrivilegeResult> {
    const c = await this._connect()
    const data = await call<Record<string, unknown>>(c, 'check_privilege_risk', { provider })
    return {
      risk: (data.risk_level as string ?? 'HIGH').toUpperCase() as PrivilegeResult['risk'],
      label: (data.risk_level as string ?? 'HIGH'),
      provider,
      indicators: (data.privilege_indicators as string[]) ?? [],
      recommendation: (data.recommendation as string) ?? '',
      notice: (data.notice as string) ?? '',
    }
  }

  async generateBriefOutline(caseType: string, facts?: string): Promise<BriefOutline> {
    const c = await this._connect()
    const data = await call<BriefOutline>(c, 'generate_brief_outline', { case_type: caseType, ...(facts ? { facts } : {}) })
    return data
  }

  async getWorkflows(): Promise<Workflow[]> {
    // Workflows come from the SKILL.md layer — no dedicated server tool.
    // Fall back to the static list from fixtures.
    const { WORKFLOWS } = await import('../fixtures')
    return WORKFLOWS as unknown as Workflow[]
  }

  async getAuditLog(): Promise<AuditEntry[]> {
    // The server's audit log is local-only. No tool exposes it remotely.
    return []
  }

  async getIntegrationStatus(): Promise<IntegrationStatus> {
    const c = await this._connect()
    const data = await call<IntegrationStatus>(c, 'integration_status', {})
    return data
  }
}

// ─── normalizers — map server JSON shapes → our TypeScript types ─────────────

function _normaliseCiteResult(citation: string, data: Record<string, unknown>): CitationResult {
  return {
    citation,
    valid: Boolean(data.valid ?? data.is_valid ?? true),
    normalized: (data.normalized ?? data.normal ?? citation) as string,
    reporter: (data.reporter ?? '') as string,
    reporter_name: (data.reporter_name ?? data.reporter ?? '') as string,
    issues: (data.issues ?? data.errors ?? []) as string[],
    integrity: (data.integrity ?? (data.in_database ? 'verified' : 'not_found')) as CitationResult['integrity'],
  }
}

function _normaliseContract(id: string, data: Record<string, unknown>): Contract {
  const rawClauses = (data.clauses as Record<string, unknown>[] | undefined) ?? []
  const clauses: ContractClause[] = rawClauses.map(cl => ({
    key: (cl.key ?? cl.clause_type ?? '') as string,
    label: (cl.label ?? cl.name ?? cl.clause_type ?? '') as string,
    text: (cl.text ?? cl.content ?? '') as string,
    risk: ((cl.risk ?? cl.risk_level ?? 'MEDIUM') as string).toUpperCase() as ContractClause['risk'],
    risk_note: (cl.risk_note ?? cl.note ?? cl.rationale ?? '') as string,
  }))

  const riskLevel = (() => {
    const hasCritical = clauses.some(c => c.risk === 'CRITICAL')
    const hasHigh = clauses.some(c => c.risk === 'HIGH')
    if (hasCritical) return 'CRITICAL'
    if (hasHigh) return 'HIGH'
    return (data.risk_level as string | undefined)?.toUpperCase() ?? 'MEDIUM'
  })() as Contract['risk_level']

  return {
    id,
    title: (data.title ?? data.contract_id ?? id) as string,
    type: (data.type ?? data.contract_type ?? 'Unknown') as string,
    risk_level: riskLevel,
    clauses,
    missing_clauses: (data.missing_clauses ?? []) as string[],
    parties: (data.parties ?? {}) as Record<string, string>,
    governing_law: (data.governing_law ?? '') as string,
    term: (data.term ?? '') as string,
    liability_cap: (data.liability_cap ?? null) as string | null,
  }
}

function _normaliseJob(raw: Record<string, unknown>): AnalysisJob {
  return {
    id: (raw.job_id ?? raw.id ?? '') as string,
    file: (raw.file_path ?? raw.file ?? '') as string,
    status: (raw.status ?? 'queued') as AnalysisJob['status'],
    queued_at: (raw.created_at ?? raw.queued_at ?? '') as string,
    completed_at: (raw.updated_at ?? raw.completed_at ?? null) as string | null,
    risk_level: (raw.risk_level ?? null) as string | null,
    clause_count: (raw.clause_count ?? null) as number | null,
    flags: (raw.flags ?? null) as number | null,
    error: (raw.error ?? undefined) as string | undefined,
  }
}

function _emptyContract(id: string): Contract {
  return { id, title: id, type: '', risk_level: 'MEDIUM', clauses: [], missing_clauses: [], parties: {}, governing_law: '', term: '', liability_cap: null }
}
